### PR TITLE
464 intel power gadget alternative for apple silicon chips

### PR DESCRIPTION
**Pull request by LuisBlanche**
- [x]  CPU Metrics with `Powermetrics`
- [x]  GPU Metrics with `PowerMetrics`
- [x]  Update doc to explain how to give `sudo` rights to `Powermetrics` (I not then fall back to constant mode)

### CODE REVIEW

**Comment by benoit-cty** [comment]
Thanks, it will be a great addition to CodeCarbon.

**Comment by LuisBlanche** [comment]
@benoit-cty this is ready for review

File: codecarbon/core/measure.py [hunk 1]
```diff
@@ -68,8 +68,23 @@ def do_measure(self) -> None:
                 self._ram_power = power
                 logger.info(
                     f"Energy consumed for RAM : {self._total_ram_energy.kWh:.6f} kWh"
-                    + f". RAM Power : {self._ram_power.W} W"
+                    + f".RAM Power : {self._ram_power.W} W"
```
Comment (benoit-cty) [inline codecarbon/core/measure.py:71]: Why removing this space after the point ? Maybe just move the point in the previous string ?

File: codecarbon/core/powermetrics.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    ):
+        self._log_file_path = os.path.join(output_dir, log_file_name)
+        self._system = sys.platform.lower()
+        self._n_points = n_points
+        self._interval = interval
+        self._setup_cli()
+
+    def _setup_cli(self):
+        """
+        Setup cli command to run Intel Power Gadget
```
Comment (benoit-cty) [inline codecarbon/core/powermetrics.py:44]:
```suggestion
        Setup cli command to run Powermetrics
```

File: codecarbon/core/powermetrics.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+                else:
+                    raise FileNotFoundError(
+                        f"Powermetrics executable not found on {self._system}"
+                    )
+        else:
+            raise SystemError("Platform not supported by Powermetrics")
+
+    def _log_values(self):
+        """
+        Logs output from Intel Power Gadget command line to a file
```
Comment (benoit-cty) [inline codecarbon/core/powermetrics.py:60]:
```suggestion
        Logs output from Powermetrics to a file
```

File: codecarbon/core/powermetrics.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+            ]
+            returncode = subprocess.call(cmd, universal_newlines=True)
+
+        else:
+            return None
+
+        if returncode != 0:
+            logger.warning(
+                "Returncode while logging power values using "
+                + f"Intel Power Gadget: {returncode}"
```
Comment (benoit-cty) [inline codecarbon/core/powermetrics.py:89]:
```suggestion
                + f"Powermetrics: {returncode}"
```

File: codecarbon/emissions_tracker.py [hunk 5]
```diff
@@ -640,7 +657,10 @@ def _do_measurements(self) -> None:
             h_time = time.time()
             # Compute last_duration again for more accuracy
             last_duration = time.time() - self._last_measured_time
-            power, energy = hardware.measure_power_and_energy(
+            (
+                power,
+                energy,
+            ) = hardware.measure_power_and_energy(  # TODO: here see how to do with Apple to split CPU / GPU
```
Comment (benoit-cty) [inline codecarbon/emissions_tracker.py:663]: The TODO has been done below, no ?

File: docs/edit/methodology.rst [hunk 6]
```diff
@@ -89,10 +89,33 @@ This mesure is not satisfying and if ever you have an idea how to enhance it ple
...
-- **On Windows or Mac**
+- **On Windows or Mac (Intel)**
 
 Tracks Intel processors energy consumption using the ``Intel Power Gadget``. You need to install it yourself from this `source <https://www.intel.com/content/www/us/en/developer/articles/tool/power-gadget.html>`_ .
 
+Apple Silicon Chips (M1, M2)
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Apple Silicon Chips contain both the CPU and the GPU.
+
+Codecarbon tracks Apple Silicon Chip energy consumption using ``powermetrics``. It should be available natively on any mac.
```
Comment (benoit-cty) [inline docs/edit/methodology.rst:100]:
```suggestion
CodeCarbon tracks Apple Silicon Chip energy consumption using ``powermetrics``. It should be available natively on any mac.
```

File: codecarbon/core/powermetrics.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+from codecarbon.core.util import detect_cpu_model
+from codecarbon.external.logger import logger
+
+
+def is_powermetrics_available():
+    try:
+        ApplePowermetrics()
+        return True
+    except Exception as e:
+        logger.debug(
```
Comment (benoit-cty) [inline codecarbon/core/powermetrics.py:19]: Do you think it's possible to detect that powermetrics exist but the user is not root to tell him how to add right ?
Like we do for RAPL: https://github.com/mlco2/codecarbon/blob/master/codecarbon/core/cpu.py#L218

**Review by benoit-cty — COMMENTED** [review]
Thanks a lot !
I've asked Amine to test on his Apple M? laptop.

File: codecarbon/emissions_tracker.py (hunk 5 above)
Comment (LuisBlanche) [reply codecarbon/emissions_tracker.py:663]: yes it's done

File: codecarbon/core/powermetrics.py (hunk 7 above)
Comment (LuisBlanche) [reply codecarbon/core/powermetrics.py:19]: I will have a look into this

**Comment by LuisBlanche** [comment]
@SaboniAmine will you have a moment to check this ?

**Review by benoit-cty — APPROVED** [review]
Thanks a lot for adding this support.

**Pull request merged**